src/library/compiler/ir_interpreter.cpp

```diff
@@ -0,0 +1,759 @@
+                        cnstr_set(o, i, eval_arg(expr_reuse_args(e)[i]));
+                    }
+                    return o;
+                }
+            }
+            case expr_kind::Proj: // object field access
+                return cnstr_get(var(expr_proj_obj(e)), expr_proj_idx(e).get_small_value());
+            case expr_kind::UProj: // USize field access
+                return box_size_t(cnstr_get_usize(var(expr_uproj_obj(e)), expr_uproj_idx(e).get_small_value()));
+            case expr_kind::SProj: { // other unboxed field access
+                size_t offset = expr_sproj_idx(e).get_small_value() * sizeof(void *) +
+                                expr_sproj_offset(e).get_small_value();
+                object *o = var(expr_sproj_obj(e));
+                switch (t) {
+                    case type::Float: throw exception("floats are not supported yet");
+                    case type::UInt8: return box(cnstr_get_uint8(o, offset));
+                    case type::UInt16: return box(cnstr_get_uint16(o, offset));
+                    case type::UInt32: return box_uint32(cnstr_get_uint32(o, offset));
+                    case type::UInt64: return box_uint64(cnstr_get_uint64(o, offset));
+                    default: throw exception("invalid instruction");
+                }
+            }
+            case expr_kind::FAp: { // satured ("full") application of top-level function
+                if (expr_fap_args(e).size()) {
+                    return call(expr_fap_fun(e), expr_fap_args(e));
+                } else {
+                    // nullary function ("constant"), cache
+                    object * const * cached = m_constant_cache.find(expr_fap_fun(e));
+                    if (cached) {
+                        return *cached;
+                    } else {
+                        object * r = load(expr_fap_fun(e), t);
+                        // the IR expects constants to be persistent
+                        // TODO(Sebastian): because of this, we currently leak these objects
+                        mark_persistent(r);
+                        m_constant_cache.insert(expr_fap_fun(e), r);
+                        return r;
+                    }
+                }
+            }
+            case expr_kind::PAp: { // unsatured (partial) application of top-level function
+                decl d = get_fdecl(expr_pap_fun(e));
+                unsigned i = 0;
+                object * cls;
+                if (void * p = lookup_symbol(expr_pap_fun(e)).m_addr) {
+                    // point closure directly to native symbol
+                    cls = alloc_closure(p, decl_params(d).size(), expr_pap_args(e).size());
+                } else {
+                    // point closure to interpreter stub taking interpreter data, declaration to be called, and partially
+                    // applied arguments
+                    // HACK: filling up closure to at least 16 arguments
```

Comment:
I hope this is a temporary hack ;)

Reply:
Yes, I will fix it after adjusting the stub for multi-threading